docs/design_notes/lift.md

```diff
@@ -0,0 +1,340 @@
+
+1. *Scope de-duplication*
+
+    This stage is only relevant if multiple Scopes are lifted together.
+    In this case we must first find the set of root scopes.
+    A scope is a root if none of its ancestors are in the set of scopes that need to be lifted.
+
+    By only lifting roots we avoid lifting the same variables twice.
+
+    For non-root scopes we store a reference to its ancestor scope and a path such that we can later reconstruct it (stage 4).
+
+2. *Filter stage*
+
+    Variables and PRNG sequences are split up into groups. This way `fn` can lift each group into the transformation separately.
+    A group is defined by a filter specified as:
+    - a list of collections/prng names
+    - `True` (match everything)
+    - `False` (match nothing)
+    - `DenyList(filter)` (match everything but the specified filter).
+
+    A collection or PRNG sequence can only be put into a single group. If a collection matches multiple filters, it will be put into the first group with a matching filter.
+    For example, `in_vars = (["params"], True)` will cause the "params" collection to be put in the first group and all other collection to be put in the second group.
+
+    For each PRNG sequence that is matched we seed a new PRNG sequence by calling `make_rng`.
+    This avoids the need to update the PRNG state after the lifted transformation is complete.
+
+3. *Transform specific lifting*
+
+    `fn` is called with the variable and PRNG groups.
+    JAX transforms have varying signatures and lifting options. Arguably the cleanest example is `vmap`.
+    In the case of vmap the function arguments, PRNGs and variable collections are passed into a `jax.vmap` wrapped function.
+
+4. *Scope reconstruction*
+
+    Now that the variables and PRNGs are lifted inside the transformation, we want to recreate the lifted scopes. Pack calls
+    `fn` with a `scope_fn` that takes the lifted variables and PRNGs and returns the reconstructed scopes with the lifted variables and rng sequences.
+
+5. *Repack stage*
+
+    After we have used the lifted scopes we have to retrieve the updated variables (PRNG sequences can simply be discarded).
+    pack passes the `repack_fn` to support this.
+    This stage is similar to stage 2 except that we only lift variables and immutable variables are ignored.
+    Immutable variables cannot be updated. Therefore, they should not be returned from the transformed function. 
+
+6. *Commit stage*
+
+    `pack` expects `fn` to return a pair where the first item will simply be returned from pack and the second item should be the repacked variables.
+    The updated variables are stored in the original/un-lifted scopes such that the mutations that happen inside the transformation survive after the transformation completes.
+
+
+### Using pack example
+
+
+A minimal example of using `pack`:
+
+```python
+from flax.core import lift
+from flax.core import Scope, init, apply, nn as core_nn
+
+def lift_id(fn, variables=True, rngs=True)
```

Comment:
While this example is very concise, it might be interesting to expand it a little bit to an example where variables/rngs are selected by filter (i.e. some are lifted and others not).

Reply:
done